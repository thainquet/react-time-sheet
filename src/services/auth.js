import axios from 'axios'

const register = (params) => {
  return axios.post(`${process.env.REACT_APP_API_URL}/register`, params);
}

const login = (params) => {
  return axios.post(`${process.env.REACT_APP_API_URL}/login`, params);
}

const changePassword = (params) => {
  return axios.post(`${process.env.REACT_APP_API_URL}/changePass`, params);
}

const forgotPassword = (params) => {
  return axios.post(`${process.env.REACT_APP_API_URL}/forgotPass`, params);
}

export default {
  login, changePassword, forgotPassword, register
}