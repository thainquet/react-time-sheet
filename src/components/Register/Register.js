
import React, { useState } from 'reactn';
import { authServices } from 'services'

const Register = (props) => {
  const { 
    initParams = { username: '', password: '', email: '' },
    initMessage = { messageUsername: '', messagePassword: '', emailMessage: '' }
 } = props;
  // handle username
  const [username, setUsername] = useState(initParams.username)
  const [messageUsername, setUsernameMessage] = useState(initMessage.messageUsername)

  const handleUsernameInput = (event) => {
    let input = event.target.value;
    let regexTV = /^[ÀÁÂÃÈÉÊÌÍÒÓÔÕÙÚĂĐĨŨƠàáâãèéêìíòóôõùúăđĩũơƯĂẠẢẤẦẨẪẬẮẰẲẴẶẸẺẼỀỀỂưăạảấầẩẫậắằẳẵặẹẻẽềềểỄỆỈỊỌỎỐỒỔỖỘỚỜỞỠỢỤỦỨỪễệỉịọỏốồổỗộớờởỡợụủứừỬỮỰỲỴÝỶỸửữựỳỵỷỹ]+$/
    let regexSpace = /\s+/g
    if (!input) {
      setUsernameMessage(messageUsername => messageUsername = "ko dc de trong")
    }
    else {
      if (input.length < 6 || regexTV.test(input) || regexSpace.test(input)) {
        setUsernameMessage(messageUsername => messageUsername = "gom 6 ky tu viet thuong tro len, viet lien ko dau")
      }
      else {
        setUsername(username => username = input)
        setUsernameMessage(messageUsername => messageUsername = '')
      }
    }
  }
  //handle password
  const [password, setPassword] = useState(initParams.password)
  const [messagePassword, setPasswordMessage] = useState(initMessage.messagePassword)

  const handlePasswordInput = event => {
    let input = event.target.value;
    if (!input) setPasswordMessage(messagePassword => messagePassword = "ko dc de trong")
    else {
      if (input.length < 8 || /^(?=.*[a-z])(?=.*[A-Z])(?=.*[0-9])/.test(input) !== true) {
        setPasswordMessage(messagePassword => messagePassword = "gom it nhat 1 chu thuong 1 chu hoa 1 chu so")
        return
      }
      else {
        setPassword(password => password = input)
        setPasswordMessage(messagePassword => messagePassword = '')
      }
    }
  }
  // handle email
  const [email, setEmail] = useState(initParams.email)
  const [emailMessage, setEmailMessage] = useState(initMessage.emailMessage)

  const handleEmailInput = event => {
    let input = event.target.value;
    let regexEmail = /^[a-z0-9_.]{5,32}@[a-z0-9]{2,}(\.[a-z0-9]{2,4}){1,2}$/
    if (!input) setEmailMessage(emailMessage => emailMessage = "ko dc de trong")
    else {
      if (regexEmail.test(input) !== true) setEmailMessage(emailMessage => emailMessage = "mail ko hop le")
      else {
        setEmail(email => email = input)
        setEmailMessage(emailMessage => emailMessage = "")
      }
    }
  }

  const handleToLoginForm = () => {
    props.history.push('/login')
  }

  const [res, resMessage] = useState('')

  const handleRegister = async () => {
    if (!username || !email || !password) {
      resMessage(res => res = "all fields required")
    }
    else 
      if (messageUsername || messagePassword || emailMessage) {
        resMessage(res => res = 'input k hop le')
        setEmail(email => email = '')
        setUsername(username => username = '')
        setPassword(password => password = '')
      }
      else{
        resMessage(res => res = '')
        const params = {
          username,
          password,
          email
        }
        const result = await authServices.register(params)
        resMessage(res => res = result.data.message)
      } 
  }

  const handlePress = event => {
    if (event.key === 'Enter') handleRegister();
  }

  return (
    <div>
      <div>
        <label htmlFor="">Username</label>
        <input type="text" autoFocus autoComplete="off" name="username" value={username} onChange={(event) => handleUsernameInput(event)} />
        <br />
        <pre>
          {messageUsername && <b>{messageUsername}</b>}
        </pre>
      </div>

      <div>
        <label htmlFor="">Password</label>
        <input type="password" autoComplete="off" name="password" value={password} onChange={(event) => handlePasswordInput(event)} />
        <br />
        <pre>
          {messagePassword && <b>{messagePassword}</b>}
        </pre>
      </div>

      <div>
        <label htmlFor="">Email</label>
        <input type="text" autoComplete="off" name="email" value={email} onChange={(event) => handleEmailInput(event)} onKeyPress={handlePress} />
        <br />
        <pre>
          {emailMessage && <b>{emailMessage}</b>}
        </pre>
      </div>

      <button onClick={() => handleRegister()} id="register-btn">Dang ky</button>
      <pre>
        {res && <b>{res}</b>}
      </pre>
      <div>
        <button onClick={() => handleToLoginForm()}>Dang nhap</button>
      </div>
    </div>
  )
}

export default Register

