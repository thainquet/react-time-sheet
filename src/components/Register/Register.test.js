import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';
import { authServices } from 'services';
import Register from './Register';

const props = {
  initParams: {
    username: 'thainq02',
    email: 'thainq00@gmail.com',
    password: 'Hihi1234'
  },
  initMessage: {
    messageUsername: '',
    messagePassword: '',
    emailMessage: ''
  },
  history: [],
  res: ''
}

it('should render correctly', () => {
  const wrapper = renderer.create(<Register {...props} />).toJSON();
  expect(wrapper).toMatchSnapshot();
});

it('has 3 inputs and 2 buttons', () => {
  const wrapper = mount(<Register {...props} />);
  expect(wrapper.find('input')).toHaveLength(3);
  expect(wrapper.find('button')).toHaveLength(2);
  wrapper.unmount();
});

it('send request for forgot password success', () => {
  const wrapper = mount(<Register {...props} />);
  const RegFunc = jest.spyOn(authServices, 'register').mockImplementation(() => Promise.resolve({ data: { code: 200, message: 'Thanh cong' } }));
  wrapper.find('button#register-btn').simulate('click')
  expect(RegFunc).toHaveBeenCalledWith(
    props.initParams
  );
  wrapper.unmount();
});
