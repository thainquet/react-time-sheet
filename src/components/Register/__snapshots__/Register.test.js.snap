// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly 1`] = `
<div>
  <div>
    <label
      htmlFor=""
    >
      Username
    </label>
    <input
      autoComplete="off"
      autoFocus={true}
      name="username"
      onChange={[Function]}
      type="text"
      value="thainq02"
    />
    <br />
    <pre>
      
    </pre>
  </div>
  <div>
    <label
      htmlFor=""
    >
      Password
    </label>
    <input
      autoComplete="off"
      name="password"
      onChange={[Function]}
      type="password"
      value="Hihi1234"
    />
    <br />
    <pre>
      
    </pre>
  </div>
  <div>
    <label
      htmlFor=""
    >
      Email
    </label>
    <input
      autoComplete="off"
      name="email"
      onChange={[Function]}
      onKeyPress={[Function]}
      type="text"
      value="thainq00@gmail.com"
    />
    <br />
    <pre>
      
    </pre>
  </div>
  <button
    id="register-btn"
    onClick={[Function]}
  >
    Dang ky
  </button>
  <pre>
    
  </pre>
  <div>
    <button
      onClick={[Function]}
    >
      Dang nhap
    </button>
  </div>
</div>
`;
